manta-trusted-setup/src/groth16/bn254/ppot.rs

```diff
@@ -0,0 +1,525 @@
+        AffineCurve, PairingEngine, SWModelParameters,
+    },
+    ff::{PrimeField, Zero},
+    pairing::Pairing,
+};
+use manta_util::into_array_unchecked;
+use memmap::{Mmap, MmapOptions};
+use std::fs::{File, OpenOptions};
+
+/// Configuration of the Perpetual Powers of Tau ceremony
+pub struct PpotCeremony;
+
+impl Size for PpotCeremony {
+    const G1_POWERS: usize = (Self::G2_POWERS << 1) - 1;
+
+    const G2_POWERS: usize = 1 << 28;
+}
+
+impl Pairing for PpotCeremony {
+    type Scalar = Fr;
+
+    type G1 = G1Affine;
+
+    type G1Prepared = <Bn254 as PairingEngine>::G1Prepared;
+
+    type G2 = G2Affine;
+
+    type G2Prepared = <Bn254 as PairingEngine>::G2Prepared;
+
+    type Pairing = Bn254;
+
+    fn g1_prime_subgroup_generator() -> Self::G1 {
+        G1Affine::prime_subgroup_generator()
+    }
+
+    fn g2_prime_subgroup_generator() -> Self::G2 {
+        G2Affine::prime_subgroup_generator()
+    }
+}
+
+impl KzgConfiguration for PpotCeremony {
+    type DomainTag = u8;
+    type Challenge = [u8; 64];
+    type Response = [u8; 64];
+    type HashToGroup = KZGBlakeHasher<Self>;
+
+    const TAU_DOMAIN_TAG: Self::DomainTag = 0;
+    const ALPHA_DOMAIN_TAG: Self::DomainTag = 1;
+    const BETA_DOMAIN_TAG: Self::DomainTag = 2;
+
+    fn hasher(domain_tag: Self::DomainTag) -> Self::HashToGroup {
+        Self::HashToGroup { domain_tag }
+    }
+
+    fn response(
+        state: &Accumulator<Self>,
+        challenge: &Self::Challenge,
+        proof: &KzgProof<Self>,
+    ) -> Self::Response {
+        let mut hasher = BlakeHasher::default();
+        for item in &state.tau_powers_g1 {
+            item.serialize_uncompressed(&mut hasher).unwrap();
+        }
+        for item in &state.tau_powers_g2 {
+            item.serialize_uncompressed(&mut hasher).unwrap();
+        }
+        for item in &state.alpha_tau_powers_g1 {
+            item.serialize_uncompressed(&mut hasher).unwrap();
+        }
+        for item in &state.beta_tau_powers_g1 {
+            item.serialize_uncompressed(&mut hasher).unwrap();
+        }
+        state.beta_g2.serialize_uncompressed(&mut hasher).unwrap();
+        hasher.0.update(&challenge);
+        proof
+            .tau
+            .serialize(&mut hasher)
+            .expect("Consuming ratio proof of tau failed.");
+        proof
+            .alpha
+            .serialize(&mut hasher)
+            .expect("Consuming ratio proof of alpha failed.");
+        proof
+            .beta
+            .serialize(&mut hasher)
+            .expect("Consuming ratio proof of beta failed.");
+        into_array_unchecked(hasher.0.finalize())
+    }
+}
+
+// impl Serializer<<Self as Pairing>::G1> for PpotCeremony {
+//     fn serialize_unchecked<W>(item: &<Self as Pairing>::G1, writer: &mut W) -> Result<(), std::io::Error>
+//     where
+//         W: ark_serialize::Write {
+// TODO : How do you distinguish between G1 and G2? They both have type sw_jacobian::GroupAffine<P>
```

Comment:
Actually there's no need to impl Serializer for PPoT.  But same question holds for Deserializer: how do I account for the fact that deserialization works differently for G1 and G2 points when these share the GroupAffine<P> type?